File: JS/sorting/merge_sort.js
// Mergesort
// O(n log(n)) time, O(n) space

function merge(arr1, arr2) {
  var mergedArr = [];

  while (arr1.length !== 0 && arr2.length !== 0) {
    if (arr1[0] < arr2[0]) {
      mergedArr.push(arr1.shift());
    } else {
      mergedArr.push(arr2.shift());
    }
  }
  return mergedArr.concat(arr1, arr2);
}

function mergeSort(arr) {
  if (arr.length <= 1) {return arr;}

  var pivotIdx = Math.floor(arr.length / 2);
  var left = arr.slice(0, pivotIdx);
  var right = arr.slice(pivotIdx);

  return merge(mergeSort(left), mergeSort(right));
}

File: Ruby/sorting/merge_sort.rb
# Mergesort
# O(n log(n)) time, O(n) space

def merge(arr1, arr2)
  merged_arr = []

  until (arr1.empty? || arr2.empty?)
    arr1[0] < arr2[0] ? merged_arr.push(arr1.shift) : merged_arr.push(arr2.shift)
  end

  merged_arr.concat(arr1).concat(arr2)
end

def merge_sort(arr)
  return arr if arr.length <= 1

  pivot_idx = arr.length / 2
  left = arr[0...pivot_idx]
  right = arr[pivot_idx...arr.length]

  merge(merge_sort(left), merge_sort(right))
end
